cpp/gdb_arrow.py

```diff
@@ -0,0 +1,1894 @@
+    gdb.Value).
+    """
+    buf = gdb.selected_inferior().read_memory(val.address, 2)
+    return struct.unpack("e", buf)[0]
+
+
+def load_atomic(val):
+    """
+    Load a std::atomic<T>'s value.
+    """
+    valty = val.type.template_argument(0)
+    # XXX This assumes std::atomic<T> has the same layout as a raw T.
+    return val.address.reinterpret_cast(valty.pointer()).dereference()
+
+
+def load_null_count(val):
+    """
+    Load a null count from a gdb.Value of an integer (either atomic or not).
+    """
+    if get_basic_type(val.type).code != gdb.TYPE_CODE_INT:
+        val = load_atomic(val)
+    return val
+
+
+def format_null_count(val):
+    """
+    Format a null count value.
+    """
+    if not isinstance(val, int):
+        null_count = int(load_null_count(val))
+    return (f"null count {null_count}" if null_count != -1
+            else "unknown null count")
+
+
+def short_time_unit(val):
+    return ['s', 'ms', 'us', 'ns'][int(val)]
+
+
+def format_month_interval(val):
+    """
+    Format a MonthInterval value.
+    """
+    return f"{int(val)}M"
+
+
+def cast_to_concrete(val, ty):
+    return (val.reference_value().reinterpret_cast(ty.reference())
+            .referenced_value())
+
+
+def scalar_class_from_type(name):
+    """
+    Given a DataTypeClass class name (such as "BooleanType"), return the
+    corresponding Scalar class name.
+    """
+    assert name.endswith("Type")
+    return name[:-4] + "Scalar"
+
+
+def array_class_from_type(name):
+    """
+    Given a DataTypeClass class name (such as "BooleanType"), return the
+    corresponding Array class name.
+    """
+    assert name.endswith("Type")
+    return name[:-4] + "Array"
+
+
+class CString:
+    """
+    A `const char*` or similar value.
+    """
+
+    def __init__(self, val):
+        self.val = val
+
+    def __bool__(self):
+        return int(data) != 0 and int(data[0]) != 0
+
+    @property
+    def data(self):
+        return self.val
+
+    def bytes_literal(self):
+        return self.val.lazy_string().value()
+
+    def string_literal(self):
+        # XXX use lazy_string() as well?
+        return string_literal(self.val)
+
+    def string(self):
+        return self.val.string()
+
+    def __format__(self, fmt):
+        return str(self.bytes_literal())
+
+
+# NOTE: gdb.parse_and_eval() is *slow* and calling it multiple times
+# may add noticeable latencies.  For standard C++ classes, we therefore
+# try to fetch their properties from libstdc++ internals (which hopefully
+# are stable), before falling back on calling the public API methods.
+
+class SharedPtr:
+    """
+    A `std::shared_ptr<T>` value.
+    """
+
+    def __init__(self, val):
+        self.val = val
+        try:
+            # libstdc++ internals
+            self._ptr = val['_M_ptr']
+        except gdb.error:
+            # fallback for other C++ standard libraries
+            self._ptr = gdb.parse_and_eval(f"{for_evaluation(val)}.get()")
```

Comment:
Would it be worth extracting a helper function like
```suggestion
        self._ptr = get_or_evaluate(lambda: val['_M_ptr'], # libstdc++
                                    lambda: val['__ptr_'], # libcxx
                                    f"{for_evaluation(val)}.get()")
```

Reply:
Hmm, as you can see in the `StdVector` constructor, there can be a bit more divergence, so I'm not sure a simple get/fallback function would suffice.